core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java

```diff
@@ -201,27 +203,51 @@
         final List<AnnotatedValueResolver> list =
                 resolver.<List<AnnotatedValueResolver>>map(ImmutableList::of).orElseGet(
                         () -> Arrays.stream(parameters)
-                                    .map(p -> of(p, pathParams, objectResolvers))
+                                    .map(p -> of(p, pathParams, objectResolvers, noAnnotationAsRequestObject))
                                     .filter(Optional::isPresent)
                                     .map(Optional::get)
                                     .collect(Collectors.toList()));
         if (list.isEmpty()) {
             throw new NoAnnotatedParameterException(constructorOrMethod.toGenericString());
         }
         if (list.size() != parameters.length) {
-            throw new NoAnnotatedParameterException("Unsupported parameter exists: " +
-                                                    constructorOrMethod.toGenericString());
+            if (list.stream().anyMatch(r -> r.annotationType() != null)) {
+                //
+                // If a user specify one of @Param, @Header or @RequestObject on the parameter list,
+                // it clearly means that the user wanted the class to be converted into a bean. e.g.
+                //
+                // class BeanA {
+                //     ...
+                //     BeanA(@Param("a") int a, int b) { ... }
+                // }
+                //
+                // But for the automatically injected types such as RequestContext and HttpRequest, etc.
+                // it is not easy to understand what a user intends for. So we ignore that.
+                //
+                // class BeanB {
+                //     ...
+                //     BeanB(ServiceRequestContext ctx, int b) { ... }
```

Comment:
Could you add a comment that this bean will not be automatically injected to clarify?

Reply:
Added [here](https://github.com/line/armeria/pull/1412/files#diff-9df84bbcae0217f0942af4e4d6a3f8c0R215). :-)